File: nvpl_lapack/ex_zgels.cpp
#include <stdio.h>
#include <nvpl_lapack.h>
#include "utils.h"

/* Main program */
int main() {
    nvpl_int_t info;
    /* Initialization */
    const nvpl_int_t m = 5;
    const nvpl_int_t n = 3;
    const nvpl_int_t nrhs = 2;
    const nvpl_int_t lda = 5;
    const nvpl_int_t ldb = 5;

    nvpl_dcomplex_t A[n][m]
            = {{1, 2, 3, 4, 5}, {1, 3, 5, 2, 4}, {1, 4, 2, 5, 3}};
    nvpl_dcomplex_t b[nrhs][m] = {{-10, 12, 14, 16, 18}, {-3, 14, 12, 16, 16}};

    printf("NVPL LAPACK version: %d\n", nvpl_lapack_get_version());

    print_zmatrix_colmajor("Entry Matrix *A", m, n, *A, lda);
    print_zmatrix_colmajor("Right Hand Side *b", m, nrhs, *b, ldb);
    printf("\n");

    printf("sgels (col-major, high-level) Example Program Results\n");
    /* Solve least squares problem*/
    char t = 'N';
    nvpl_dcomplex_t *work = NULL;
    nvpl_dcomplex_t w;
    nvpl_int_t lwork = -1;
    NVPL_LAPACK_zgels(&t, &m, &n, &nrhs, *A, &lda, *b, &ldb, &w, &lwork, &info);
    lwork = (nvpl_int_t)nvpl_dcomplex_t_real(w);
    printf("lwork = %ld", (long int)lwork);
    work = (nvpl_dcomplex_t *)malloc(sizeof(nvpl_dcomplex_t) * lwork);
    if (!work) {
        printf("insufficient memory");
        exit(-1);
    }
    NVPL_LAPACK_zgels(
            &t, &m, &n, &nrhs, *A, &lda, *b, &ldb, work, &lwork, &info);

    /* Print Solution */
    print_zmatrix_colmajor("Solution", n, nrhs, *b, ldb);
    printf("\n");

    free(work);

    exit(info);
}

File: nvpl_lapack/f77/ex_zgels.f
program usageexample
    implicit none
    double complex, allocatable :: a(:, :)
    double complex, allocatable :: b(:, :)
    double complex, allocatable :: work(:)
    double complex :: work_query(1)
    external zgels

    character(1) trans
    integer m, n, nrhs
    integer lda, ldb, lwork
    integer info
    
    ! Problem dimensions
    m = 5      ! Number of rows of A and b matrices
    n = 3      ! Number of columns of A
    nrhs = 2   ! Number of columns of b
    lda = m    ! Leading dimension of A
    ldb = m    ! Leading dimension of b
    trans = 'N'
    info = 0
    
    ! Allocate matrices
    allocate(a(lda, n))
    allocate(b(ldb, nrhs))
    
    ! Initialize Matrix A (5x3): column-major storage
    ! A = 1   1   1
    !     2   3   4  
    !     3   5   2
    !     4   2   5
    !     5   4   3
    a = reshape([dcmplx(1.0d0,0.0d0), dcmplx(2.0d0,0.0d0), dcmplx(3.0d0,0.0d0), &
                 dcmplx(4.0d0,0.0d0), dcmplx(5.0d0,0.0d0), dcmplx(1.0d0,0.0d0), &
                 dcmplx(3.0d0,0.0d0), dcmplx(5.0d0,0.0d0), dcmplx(2.0d0,0.0d0), &
                 dcmplx(4.0d0,0.0d0), dcmplx(1.0d0,0.0d0), dcmplx(4.0d0,0.0d0), &
                 dcmplx(2.0d0,0.0d0), dcmplx(5.0d0,0.0d0), dcmplx(3.0d0,0.0d0)], [lda, n])

    ! Initialize Matrix b (5x2): column-major storage
    ! b = -10  -3
    !      12  14
    !      14  12
    !      16  16
    !      18  16
    b = reshape([dcmplx(-10.0d0,0.0d0), dcmplx(12.0d0,0.0d0), &
                 dcmplx(14.0d0,0.0d0), dcmplx(16.0d0,0.0d0), &
                 dcmplx(18.0d0,0.0d0), dcmplx(-3.0d0,0.0d0), &
                 dcmplx(14.0d0,0.0d0), dcmplx(12.0d0,0.0d0), &
                 dcmplx(16.0d0,0.0d0), dcmplx(16.0d0,0.0d0)], [ldb, nrhs])
    
    ! Print input matrices
    call print_zmatrix(a, m, n, lda, 'Entry Matrix A')
    call print_zmatrix(b, m, nrhs, ldb, 'Right Hand Side Matrix b')
    
    print *, "zgels Example Program Results"
    print *
    
    ! Workspace query
    lwork = -1
    call zgels(trans, m, n, nrhs, a, lda, b, ldb, work_query, lwork, info)
    lwork = int(real(work_query(1)))
    
    print *, "Optimal workspace size: ", lwork
    
    ! Allocate workspace
    allocate(work(lwork))
    
    ! Solve least squares problem: min_x || A * x - b ||
    call zgels(trans, m, n, nrhs, a, lda, b, ldb, work, lwork, info)
    
    ! Any errors?
    if (info < 0) then
        print *, "Illegal input argument."
    else if (info > 0) then
        print *, "Matrix A doesn't have a full rank."
    end if
    
    if (info .ne. 0) then
        deallocate(a, b, work)
        stop info
    end if
    
    print *, "Solution (first n rows of b):"
    call print_zmatrix(b, n, nrhs, ldb, 'x')
    
    ! Clean up
    deallocate(a, b, work)
    
end program usageexample
